### PR TITLE
[ui] General keyboard navigation in the UI

### LINKED ISSUE #12762: UI: general keyboard navigation
It would be nice to use keyboard commands to navigate the Nomad UI.

I'd like to move between nav/subnav sections with arrow key modifiers, or vim-like go-to keys.

## Launch Blockers
- [x] Shift-key vimium-style "hint" highlighting
- [x] #13046
- [x] #13777
- [x] Extend Custom Modifier to to allow hints on pre-loaded items (main menu)
- [x] #13890
- [x] Shortcuts modal: add dismiss button; trap focus
- [x] #13477
- [x] #13753
- [x] Fix shiftFriendlyBuffer to allow up/down arrow keys to work (currently auto-trigger shift commands)
- [x] Bugfix: "u" to go up a level in jobs has an extra stage

## Later:
- [ ] #13478
- [ ] #13495
- [ ] Handle table sorting for enumerable keyboard nav items
- [ ] Add `Shift + ->` etc. hint to subnav and nav

### PR DESCRIPTION
**Pull request by philrenaud**
Resolves #12762 

## A general Keyboard / Keynav service for the Nomad UI
- This allows Nomad users to navigate and operate their UI by use of keyboard alone and provides powerful defaults to move around:
  - absolute routes (such as `g j` to go to jobs index),
  - relative routes (such as `shift + right-arrow` to move to the next subnav section), and
  - dynamic routes (such as `shift + 03` to go to the third link in a given list)
- Creates a new global modal, accessible by pressing `?` anyplace in the app
- Gives users the choice to enable/disable shortcuts, or more advanced, to rebind individual commands to the pattern of their choosing.

## Implementation and extension
- Establishes a `keyboard` service that emits a consumable `buffer` array and `matchedCommand` object
- Listens for keystrokes in any non-input elements
- On each keystroke, if the buffer matches a pattern/sequence in `keyboard.keyCommands`, that command's `action` is run. These actions can be any function, and are commonly `router.transitionTo()` objects.
- Uses [Tether](https://www.npmjs.com/package/tether) to place hints on-screen for when the user holds `Shift`
- Shortcuts can be added by use of modifier (such as [the Go to ACLs link](https://github.com/hashicorp/nomad/pull/12831/files#diff-1ec2d76a07ab84748c05e126dc5a4acd5cc4ff74274d8e7c2232bb46274ea887R63)), or by use of component-level, conditional helper `keyboard-commands` (for example, how we [conditionally add the escape key to Evaluations modal](https://github.com/hashicorp/nomad/pull/12831/files#diff-bc34608303f7472092dae6e0956c0751561489b44098aae229b22be72fb6b94dR2-R4))
- Buffer clears after 0.75sec or when a successful keyboard command is executed

## Side-effects
- In order to facilitate a `u` shortcut to "Go up a level" from nested routes, reworked many otherwise-inline subnavs into their own components, such as the [plugin subnav](https://github.com/hashicorp/nomad/pull/12831/files#diff-ed282c6370801ee51d43fbae8b5f449ee01b3c71e31a6ac1ac3c333c1c587674R1-R6)
- In order to work around an Ember Router bug around LinkTos with query properties when deeply nested, invoked [the classic linkTo component](https://github.com/hashicorp/nomad/pull/12831/files#diff-e59da3ad52b5b03e7a6fdb11e2402d729cd46a998234a63bc409166522c8655bR1-R13) sparingly

*Showing the keyboard modal (user pressed "?"):*
image

*Showing hints active (user holding "Shift"):*
image

### CODE REVIEW

File: ui/app/transitions.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+export default function () {
+  // Add your transitions here, like:
+  //   this.transition(
+  //     this.fromRoute('people.index'),
+  //     this.toRoute('people.detail'),
+  //     this.use('toLeft'),
+  //     this.reverse('toRight')
+  //   );
```
Comment (ChaiWithJai) [inline ui/app/transitions.js:8]: `blocking`: Are you sure you want to commit this?

File: ui/package.json [hunk 2]
```diff
@@ -97,7 +98,7 @@
     "ember-inline-svg": "^1.0.1",
     "ember-load-initializers": "^2.1.2",
     "ember-maybe-import-regenerator": "^1.0.0",
-    "ember-modifier": "^3.1.0",
+    "ember-modifier": "3.2.7",
```
Comment (ChaiWithJai) [inline ui/package.json:101]: `question`:  `3.2` was a major change to this [library](https://github.com/ember-modifier/ember-modifier/releases/tag/v3.2.0). Did you check the changelog and ensure we won't have any breaking changes (either from us or dependencies)?

File: ui/tests/acceptance/keyboard-test.js [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+      await triggerKeyEvent('.page-layout', 'keydown', 'ArrowRight', {
+        shiftKey: true,
+      });
+      assert.equal(
+        currentURL(),
+        `/jobs/${jobID}@default`,
+        'Shift+ArrowRight takes you to the first tab in the loop'
+      );
+
+      // await this.pauseTest();
```
Comment (ChaiWithJai) [inline ui/tests/acceptance/keyboard-test.js:349]: `blocking`: Do you want to commit this?

**Review by ChaiWithJai — APPROVED** [review]
Spotted a few lurking bugs and code that shouldn't have been committed so just be careful around that.

File: ui/package.json (hunk 2 above)
Comment (ChaiWithJai) [reply ui/package.json:101]: `blocking`: I think we need to drop support for `node: 12` as a result.

I get the following error message on build:

`The engine "node" is incompatible with this module. Expected version ">=14". Got "12.22.1"`.

We should update the `node` version in the `engines` stanza here to alert our users.

cc:  @DingoEatingFuzz

File: ui/app/components/breadcrumbs/default.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  @service router;
+
+  @action
+  traverseUpALevel(args) {
+    const [path, ...rest] = args;
+    this.router.transitionTo(path, ...rest);
+  }
+
+  get maybeKeyboardShortcut() {
+    return this.args.isOneCrumbUp() ? KeyboardShortcutModifier : null;
```
[the remark this answers is not in the document]
Comment (philrenaud) [reply ui/app/components/breadcrumbs/default.js:16]: isOneCrumbUp() has its parameters passed as an `(action)` in app-breadcrumbs.hbs — it template-parameterizes the total number of breadcrumbs and its current position. This is a closure meant to keep singleton state isolated.

File: ui/app/transitions.js (hunk 1 above)
Comment (philrenaud) [reply ui/app/transitions.js:8]: Good catch; this was an early abandoned idea

File: ui/package.json (hunk 2 above)
Comment (philrenaud) [reply ui/package.json:101]: Tested against our resize, code-mirror, and autofocus existing modifiers. I think we're good here; we'll need to do some migrations before 4.0.

**Comment by philrenaud** [comment]
Closing; was split into https://github.com/hashicorp/nomad/pull/14138 and https://github.com/hashicorp/nomad/pull/14165

**Pull request closed without merging**